deployment/terraform/collect.go

```diff
@@ -0,0 +1,239 @@
+		sshc, err := extAgent.NewClient(agent.PublicIP)
+		if err != nil {
+			return fmt.Errorf("error in getting ssh connection %w", err)
+		}
+		clients[fmt.Sprintf("agent%d", i)] = sshc
+		if i == 0 {
+			clients["coordinator"] = sshc
+		}
+	}
+
+	var collection []collectInfo
+	addInfo := func(instance, src string, compress bool) {
+		collection = append(collection, collectInfo{
+			instance,
+			src,
+			compress,
+		})
+	}
+	for name := range clients {
+		switch {
+		case name == "proxy":
+			addInfo(name, "/var/log/nginx/error.log", true)
+			addInfo(name, "/etc/nginx/nginx.conf", false)
+			addInfo(name, "/etc/nginx/sites-enabled/mattermost", false)
+		case strings.HasPrefix(name, "app"):
+			addInfo(name, "/opt/mattermost/logs/mattermost.log", true)
+			addInfo(name, "/opt/mattermost/config/config.json", false)
+		case strings.HasPrefix(name, "agent"):
+			addInfo(name, "/home/ubuntu/agent.log", true)
+		case name == "coordinator":
+			addInfo(name, "/home/ubuntu/mattermost-load-test-ng/ltcoordinator.log", true)
+			addInfo(name, "/home/ubuntu/mattermost-load-test-ng/config/config.json", false)
+			addInfo(name, "/home/ubuntu/mattermost-load-test-ng/config/coordinator.json", false)
+			addInfo(name, "/home/ubuntu/mattermost-load-test-ng/config/simplecontroller.json", false)
+			addInfo(name, "/home/ubuntu/mattermost-load-test-ng/config/simulcontroller.json", false)
+			continue
+		}
+		addInfo(name, "dmesg", false)
+	}
+
+	var wg sync.WaitGroup
+	filesChan := make(chan file, len(collection))
+	wg.Add(len(collection))
+	for _, info := range collection {
+		go func(info collectInfo) {
+			defer wg.Done()
+
+			sshc := clients[info.instance]
+
+			var downloadPath string
+
+			if !filepath.IsAbs(info.src) {
+				cmd := info.src
+				info.src = fmt.Sprintf("/tmp/%s.log", info.src)
+				cmd = fmt.Sprintf("%s > %s", cmd, info.src)
+				if _, err := sshc.RunCommand(cmd); err != nil {
+					fmt.Printf("failed to run cmd %q: %s\n", cmd, err)
+					return
+				}
+			}
+
+			if info.compress {
+				downloadPath = fmt.Sprintf("/tmp/%s.xz", filepath.Base(info.src))
+				cmd := fmt.Sprintf("cat %s | xz -2 -T4 > %s", info.src, downloadPath)
+				if _, err := sshc.RunCommand(cmd); err != nil {
+					fmt.Printf("failed to run cmd %q: %s\n", cmd, err)
+					return
+				}
+			}
+
+			if downloadPath == "" {
+				downloadPath = info.src
+			}
+
+			var b bytes.Buffer
+			if err := sshc.Download(downloadPath, &b, false); err != nil {
+				fmt.Printf("failed to download file %q: %s\n", downloadPath, err)
+				return
+			}
+
+			fmt.Printf("collected %s from %s instance\n", filepath.Base(downloadPath), info.instance)
+
+			file := file{
+				name: fmt.Sprintf("%s_%s", info.instance, filepath.Base(downloadPath)),
+				data: b.Bytes(),
+			}
+
+			filesChan <- file
+		}(info)
+	}
+
+	wg.Wait()
+
+	numFiles := len(filesChan)
+	if numFiles == 0 {
+		return errors.New("failed to collect any file")
+	}
+
+	files := make([]file, numFiles)
```

Comment:
mmm .. what if some files succeed and some fail. Then won't you be waiting forever ? Perhaps it might be better to just range over the channel in a separate goroutine and close it after `wg.Wait` is done to let it auto-exit. Then I believe we don't even need it to be buffered.

Reply:
If some files fail to get collected they won't get sent into the channel. We only loop on the actual content (`numFiles`) that we find in the channel.

Reply:
Ah missed that. Thanks.